File: NetInterface/src/sk/hackcraft/util/PeriodicWorkDispatcher.java
package sk.hackcraft.util;


public abstract class PeriodicWorkDispatcher
{
	private MessageQueue messageQueue;
	private long pause;
	
	private boolean active;
	
	public PeriodicWorkDispatcher(MessageQueue messageQueue, long pause)
	{
		this.messageQueue = messageQueue;
		this.pause = pause;
		this.active = false;
	}
	
	public void start()
	{
		if (active)
		{
			return;
		}
		
		active = true;
		
		messageQueue.post(new Runnable()
		{
			@Override
			public void run()
			{
				if (active)
				{
					doWork();
					messageQueue.postDelayed(this, pause);
				}
			}
		});
	}
	
	public void stop()
	{
		if (!active)
		{
			return;
		}
		
		active = false;
	}
	
	protected abstract void doWork();
}

File: CSharpNetInterface/Util/PeriodicWorkDispatcher.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace CSharpNetInterface.Util
{
	public class PeriodicWorkDispatcher
	{
		private readonly IMessageQueue messageQueue;
		private readonly long pauseDelay;
		private readonly Action work;
	
		private bool active;
	
		public PeriodicWorkDispatcher(IMessageQueue messageQueue, long pauseDelay, Action work)
		{
			this.messageQueue = messageQueue;
			this.pauseDelay = pauseDelay;
			this.work = work;

			this.active = false;
		}
	
		public void Start()
		{
			if (active)
			{
				return;
			}
		
			active = true;

			messageQueue.Post(() => Work());
		}

		private void Work()
		{
			if (active)
			{
				work();
				messageQueue.PostDelayed(() => Work(), pauseDelay);
			}
		}
	
		public void Stop()
		{
			if (!active)
			{
				return;
			}
		
			active = false;
		}
	}
}
